lib/ui/window.dart

```diff
@@ -143,14 +161,234 @@ class Locale {
   /// ```
   ///
   /// The primary language subtag must not be null. The region subtag is
-  /// optional.
-  ///
-  /// The values are _case sensitive_, and should match the case of the relevant
-  /// subtags in the [IANA Language Subtag
-  /// Registry](https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry).
-  /// Typically this means the primary language subtag should be lowercase and
-  /// the region subtag should be uppercase.
-  const Locale(this._languageCode, [ this._countryCode ]) : assert(_languageCode != null);
+  /// optional. The values are _case sensitive_, and must match the values
+  /// listed in
+  /// http://unicode.org/repos/cldr/tags/latest/common/validity/language.xml and
+  /// http://unicode.org/repos/cldr/tags/latest/common/validity/region.xml.
+  ///
+  /// This method only produces standards-compliant instances if valid language
+  /// and country codes are provided. Deprecated subtags will be replaced, but
+  /// incorrectly cased strings are not corrected.
+  const Locale(
+    this._languageCode, [
+    this._countryCode,
+  ]) : assert(_languageCode != null),
+       this._scriptCode = null,
+       this._variants = null,
+       this._extensions = null;
+
+  /// Creates a new Locale object with the specified parts.
+  ///
+  /// This is for internal use only. All fields must already be normalized, must
+  /// already be canonicalized. This method does not modify parameters in any
+  /// way or do any syntax checking.
+  ///
+  /// * language, script and region must be in canonical form.
+  /// * iterating over variants must provide variants in alphabetical order.
+  /// * The extensions map must contain only valid key/value pairs. "u" and "t"
+  ///   keys must be present, with an empty string as value, if there are any
+  ///   subtags for those singletons.
+  Locale._internal(
+    String language, {
+    String script,
+    String region,
+    collection.SplayTreeSet<String> variants,
+    collection.SplayTreeMap<String, String> extensions,
+  }) : assert(language != null),
+       assert(language.length >= 2),
+       assert(language.length <= 8),
+       assert(language.length != 4),
+       assert((script ?? 'Xxxx').length == 4),
+       assert((region ?? 'XX').length >= 2),
+       assert((region ?? 'XX').length <= 3),
+       _languageCode = language,
+       _scriptCode = script,
+       _countryCode = region,
+       _variants = collection.LinkedHashSet<String>.from(variants ?? <String>[]),
```

Comment:
if we want this as a LinkedHashSet, why do we not accept one as an argument?

Reply:
I'm adjusting the documentation:

  /// * Iterating over variants must provide variants in alphabetical order. We
  ///   force this by taking a SplayTreeSet as parameter input, which
  ///   automatically sorts its items. We then convert to a LinkedHashSet for
  ///   faster access and iteration.

How does that sound? It could be considered premature optimization. Maybe we should just store the SplayTreeSet, when it's provided. How does that sound? (Variants are very rare, multiple variants probably doubly so. We could consider storing a String and split on demand even, which is what ICU / other languages seem to do.)

Reply:
Why not just take a LinkedHashSet or an Iterable? It's very weird to specifically ask for a SplayTreeSet, that's a very specialized type (to the point where IIRC there was talk of taking it out of the standard library).

Reply:
Ah: so _internal has to use initializer lists. Conceptually, I want to do:
`_variants = sorted(variants)`
This way, _internal can guarantee/control the structure of internal fields, even if it doesn't do subtag validation. Using a SplayTreeSet as parameter is one way to make this happen. Alternatively I would have to write (or find) a function I can call. Or I could create another internal/private factory constructor that does the list sorting and then calls the _internal() constructor.

Reply:
https://github.com/flutter/engine/pull/6518 is going with a List<String>.